plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -190,27 +188,39 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
             t' <- processTerm t
             -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might
             -- actually have got rid of all of them!
-            pure $ mkLet a NonRec bs' t'
+            pure $ mkLet ann NonRec bs' t'
         -- This includes recursive let terms, we don't even consider inlining them at the moment
-        t -> forMOf termSubterms t processTerm
-    applyTypeSubstitution :: Type tyname uni a -> InlineM tyname name uni fun a (Type tyname uni a)
-    applyTypeSubstitution t = gets isTypeSubstEmpty >>= \case
-        -- The type substitution is very often empty, and there are lots of types in the program,
-        -- so this saves a lot of work (determined from profiling)
-        True -> pure t
-        _    -> typeSubstTyNamesM substTyName t
-    -- See Note [Renaming strategy]
-    substTyName :: tyname -> InlineM tyname name uni fun a (Maybe (Type tyname uni a))
-    substTyName tyname = gets (lookupType tyname) >>= traverse liftDupable
-    -- See Note [Renaming strategy]
-    substName :: name -> InlineM tyname name uni fun a (Maybe (Term tyname name uni fun a))
-    substName name = gets (lookupTerm name) >>= traverse renameTerm
-    -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-    -- Already processed term, just rename and put it in, don't do any further optimization here.
-    renameTerm ::
-        InlineTerm tyname name uni fun a
-        -> InlineM tyname name uni fun a (Term tyname name uni fun a)
-    renameTerm (Done t) = liftDupable t
+        t -> do
+            processedT <- forMOf termSubterms t processTerm
```

Comment:
Also it's not generally quadratic! We only recurse into subterms once: here. So it's traversing everything once in the main loop.

What _is_ slightly risky is that `inlineSat` does a traversal looking downwards to see if an application is saturated. And because we process subterms first, we will do this for each prefix of the application first. So the runtime _is_ quadratic in the length of the largest saturated applications, but we just have to live with that, I think. Absolutely worth a comment, though.

e.g. what happens in this program?
```
let x = \a1 ... an . 1
in x 1 2 ... n
```

Also worth noting that this _won't_ happen in the context-based version (why?), which is a disadvantage of our approach.